### ISSUE TITLE
Aggregation dimension classifications

### ISSUE THREAD

**Issue by jgoizueta** [CartoDB/Windshaft-cartodb#1037]
We need to add `dimensions` control to the aggregation API, so that, in addition to the spatial aggregation (controlled by `resolution` and `placement`) we can add additional expressions into the GROUP BY of the aggregation.

We don't want to allow arbitrary SQL expressions for the dimensions. We'd like to support categorical columns and classified continues values. For the latter, they could be numbers or times and we need to control the granularity of the classification, e.g. if the time is shown as hours, days, months... or if prices are shown in 3 intervals, etc.

We'll start by implementing the time case and validate the API by [integrating its use in CARTO VL](https://github.com/CartoDB/carto-vl/issues/551)

For time aggregation, there are two interesting cases:
* Grouping a time series into time units such as minutes, hours, days, months, years. A refinement of this would be two allow groups of N units, e.g. 10-minute buckets, 4-month, etc.
* Grouping as before into time units, but computing them relative to (modulo) a larger time unit: hours of the day (0-23), days of the week (0-6), months of the year (1-12), ...

### API parameters

We could give support to these needs with an API like this: (`dimensions` would be a parameter within the `aggregation` of the layer)

```javascript
dimensions: {
  hour_of_day: {
    column: 'time_of_arrival'
    granularity: 'hour',
    multiplicity: 1, // optional: number of units per grain
    cycle: 'day',    // optional: granularity relative to this cycle
    offset: 0,       // optional: starting value for granularity
    timezone: -7200  // optional: UTC offset in seconds of the result
  }
}
```

Note: this nomenclature is unfortunate, because the date histogram already uses `offset` for the time zone offset. So we must review this. (if for example we decide not to support the offset defined here, see below, be could use  `offset` for the timezone offset for consistency with histograms).

The offset parameter could be useful to specify both the order of cyclic units and how they are grouped:
* With multiplicity > 1, e.g. 3 for months/year, by default we would get quarters as jan-feb-mar, may-apr-jun, etc.
  If we use an offset of 1, we would get feb-mar-may, ... nov-dec-jan instead.
* With multiplicity = 1 it would simply define the starting point of the cycle, e.g. what day of the week is the first (0), sunday, monday?

The timezone parameter is necessary because it determines how time is broken up into days, etc. The user would specify in which timezone is the classification to be performed.

The general SQL expression for the classification would be equivalent to:
(with `usec` the duration in seconds of the grain units, `csec` the duration of the cycle in seconds, `m` the multiplicity, `tz` the timezone)

* Cyclic case:
  `(FLOOR(date_part('epoch', timezone(tz, column)) / (usec*m)) - offset) % (csec/(m*usec))`
* Non-cyclic case:
  `FLOOR(ate_part('epoch', timezone(tz, column) / (usec*m)) - offset`

This is just a conceptual example, it would work only for regular units of time, but months and larger units are irregular because their duration in seconds is not fixed.

> If we're really into offsets we could add an additional cyclic_offset so that both the start of groupings of multiple units, and the order of the resulting grouping order (for cycles) can be independently defined. The resulting expression would be:
`(FLOOR(date_part('epoch', timezone(tz, column)) / (usec*m)) - offset - c_offset*csec/(m*usec)) % (csec/(m*usec))`
I hope we all agree we don't need that much.

### Result Units

Then we need to define what units to use for the result.
To support clients through MVT it seems best to limit ourselves to numeric formats (not dates or strings). Then we have these options:
* For the cyclic cases we have some common cases that users will expect:
  + hour of day: 0-23
  + day of month: 1-31
  * month of year 1-12
  * day of week: 0-6 or 1-7 ?
* For the non-cyclic case I see three options:
  + ordinal number for the specify grain (as in the SQL expression above); e.g. if the grain is three-month periods,
    then if 2018-01-01 corresponds to N, 2018-03-01 will correspond to N+1.
  + ordinal numbers for the specify granularity unit (we'd need to multiply the SQL expression by the multiplicity);
    then if 2018-01-01 corresponds to N, 2018-03-01 will correspond to N+3 whatever the multiplicity is.
  + UNIX epoch number in seconds (so we generate full dates truncated to the grain precision)

For the last case the advantage is that the values could be easily formatted as dates, compared with other dates, etc, and we would have homogenous values independently of the units used.
But the value spacing would be irregular. (this irregalarity could be beneficial if we want to animate dates, but in general, e.g. for ramp-coloring, charts, having the ordinal values would be more useful and consistent with the cyclic case)

#### Group data by day of week, with sunday=0, etc.

```javascript
dimensions: {
  dow: {
    column: 'the_date'
    granularity: 'day',
    cycle: 'week'
  }
}
```

Resulting SQL expression: `date_part('dow', the_date)`

#### Group data by day of week, with monday=0, etc.

```javascript
dimensions: {
  dow: {
    column: 'the_date'
    granularity: 'day',
    cycle: 'week',
    offset: 1
  }
}
```

Resulting SQL expression: `(date_part('dow', the_date) - 1) % 7`

Note: we could also use the ISO convention 1=monday, etc and use `date_part('isodow', the_date)`

#### Group data by hour of day (0-23)

```javascript
dimensions: {
  hour: {
    column: 'the_date'
    granularity: 'hour',
    cycle: 'day'
  }
}
```

Resulting SQL expression: `date_part('hour', the_date)`

#### Group data by days

```javascript
dimensions: {
  interval: {
    column: 'the_date'
    granularity: 'month'
  }
}
```

SQL: `date_part('epoch', date_trunc('day', the_date))/(24*60*60)`

#### Group data by months

```javascript
dimensions: {
  interval: {
    column: 'the_date'
    granularity: 'month'
  }
}
```

SQL: `date_part('month', the_date) + 12*(date_part('year', the_date)-date_part('year', to_timestamp(0.0)))`

#### Group data by 10-minute intervals

```javascript
dimensions: {
  interval: {
    column: 'the_date'
    granularity: 'minute',
    multiplicity: 10
  }
}
```

SQL: `FLOOR(date_part('epoch', date_trunc('day', the_date))/60 / 10.0)`,

### Questions

* Units of the result. I say: let's go for the first option above, i.e. ordinal number of groupings, be it 10-minute groupings or 3-month or whatever
* Do we need offset?. I'd say let's do it. So maybe we could rename `offset` to start and `timezone` to `offset` or perhaps `tz_offset`.
* What to do with uncommon cyclic cases.
We'll use specific expressions for the cases such as month-of-year (`date_part('month', the_date)`). For unexpected combinations of units
(e.g. minute/day), we could either not support it (produce error) or use a generic SQL expression such as the above. I'd leave the generic expression as a fall-back to experiment with it.
* Are you happy with the parameter names? (I am OK with them).

cc @dgaubert @rochoa @davidmanzanares @Algunenano  I'd like you to take a look at this and specifically at the questions, to see if you're answers differ from mine.

**Comment by Algunenano** [comment]
TLDR: Let's start supporting what we can do natively with Postgresql functions (`date_part`, `extract`) and wait to see if supporting other subfields / cycles is really necessary.

I feel this is a super complex approach that is trying to handle too many options at once, which is not wrong, but I think we should wait to see if all those options are really needed, specially the `multiplicity`, `cycle` and `offset` parameter. Adding and supporting all these options will make this a big monster.
 
>    timezone: -7200  // optional: UTC offset in seconds of the result

After working in the histograms I can say that supporting this as numeric is a huge PITA and doing it correctly is literally impossible (you can't handle timezone switches with just the offset from UTC). I would strongly recommend moving to a string based timezone, supported and handled by Postgresql.

> This is just a conceptual example, it would work only for regular units of time, but months and larger units are irregular because their duration in seconds is not fixed.

Not really; hours and days are also irregular, you can have leap seconds, changes in timezones (to half hours for example) and so on.

> Then we need to define what units to use for the result.
To support clients through MVT it seems best to limit ourselves to numeric formats (not dates or strings). Then we have these options:

I'd avoid strings not only for MVT, but for the simple reason that they come in a certain language. Epoch is universal and the frontend can do whatever it wants with it.

Let's start with whatever is more simple using native functions and avoid adding math that comes back to bite ue:

Example for grouping by day (with start as 00:00):
```sql
Select date, min(cartodb_id) as cartodb_id from
(
    Select date_trunc('day', timezone('Europe/Madrid', time)) as date, cartodb_id FROM all_month
) _a
group by date order by date
limit 5;
        date         | cartodb_id 
---------------------+------------
 2017-07-11 00:00:00 |      10492
 2017-07-12 00:00:00 |      10130
 2017-07-13 00:00:00 |       9621
 2017-07-14 00:00:00 |       9202
 2017-07-15 00:00:00 |       8796
(5 rows)

LIMIT 5;
```


Grouping by week of the year:
```sql
Select date, min(cartodb_id) as cartodb_id from
(
    Select date_part('week', timezone('Europe/Madrid', time)) as date, cartodb_id FROM all_month
) _a
group by date order by date;
 date | cartodb_id 
------+------------
   28 |       8373
   29 |       5648
   30 |       3172
   31 |        899
   32 |          1
(5 rows)
```

By day of the week (0..6. Use isodow for 1..7)
```sql
Select date, min(cartodb_id) as cartodb_id from
(
    Select extract('dow' from timezone('Europe/Madrid', time)) as date, cartodb_id FROM all_month
) _a
group by date;
```

> For the last case the advantage is that the values could be easily formatted as dates, compared with other dates, etc, and we would have homogenous values independently of the units used.
But the value spacing would be irregular. (this irregalarity could be beneficial if we want to animate dates, but in general, e.g. for ramp-coloring, charts, having the ordinal values would be more useful and consistent with the cyclic case)

You are asking for irregular formats, you get irregular ranges. 

> Units of the result. I say: let's go for the first option above, i.e. ordinal number of groupings, be it 10-minute groupings or 3-month or whatever

IMO, I would offer epoch output for everything and, for known ranges (hours of the day, days of the week or month, etc) I would also offer the appropriate ordinal part (0..23 for hours of the day or 0 of 6 for day of the week). Trying to use ordinals for irregular intervals where it's not handled by Postgresql will be a headache (it can be done, but expect a lot of corner cases).

> Do we need offset?. I'd say let's do it. So maybe we could rename offset to start and timezone to offset or perhaps tz_offset.

I don't think it's worth it right now. I don't see anyone really needing the hours to start at minute 05 instead of 00, for example.

**Comment by jgoizueta** [comment]
@Algunenano, regarding this:

> After working in the histograms I can say that supporting this as numeric is a huge PITA and doing it correctly is literally impossible (you can't handle timezone switches with just the offset from UTC). I would strongly recommend moving to a string based timezone, supported and handled by Postgresql.

I saw [your code](https://github.com/CartoDB/Windshaft-cartodb/blob/6ba9e50da741e8c012d8eb30ca3fb2071b2845b0/lib/cartodb/models/dataview/histograms/date-histogram.js#L6-L27) to handle time zones in the histograms, but  what I don't understand is, I've been using expressions of the kind `timezone('${tz} seconds', ${t})\`  where tz is the offset in seconds and t is a date column and they worked for me. Isn't this deterministic? can this yield different results depending on the kind of the date column or the PG version or any other thing?

**Comment by davidmanzanares** [comment]
My 2 cents: I think we can remove a little bit of functionality, but I think that keeping `cycle` would be better (IMO, this is a common case).

Regarding the return type, I agree on returning Epoch in general, but in combination with cycle something needs to be done since `Monday` or `13:45` is not a particular point in time after Epoch.

**Comment by Algunenano** [comment]
> I saw your code to handle time zones in the histograms, but what I don't understand is, I've been using expressions of the kind timezone(INTERVAL '${tz} seconds', ${t})\ where tz is the offset in seconds and t is a date column and they worked for me. Isn't this deterministic? can this yield different results depending on the kind of the date column or the PG version or any other thing?

Nobody needs offset `+3600`. They need the time in `Madrid` which sometimes is GMT +1 and some other times is GMT+2 so requesting `3600` is only right half the time (and force us to do funny translations for Postgresql).

**Comment by jgoizueta** [comment]
@dgaubert: I intended `granularity` to be a single simple time unit for simplicity: second, minute, hour, day, etc. Then to support buckets such as 6 months (a semester) or 10 minutes (which could be interesting in  some data-logging situations) or 15 days, etc I added the `multiplicity`, which might not be a very good name, it's just the number of units. 

On second thought the names are confusing, maybe we could have `granularity_units`and `number_or_units` or something.

To have a single `granularity` property it would have to accept expressions such as "6 months". I prefer to have separate properties to avoid the parsing.

**Comment by jgoizueta** [comment]
> Nobody needs offset +3600. They need the time in Madrid which sometimes is GMT +1 and some other times is GMT+2 so requesting 3600 is only right half the time (and force us to do funny translations for Postgresql).

Very true @Algunenano. I'll think about either allowing any name supported by PG.

By the way, regarding all your other concerns about complexity of this... My approach now is to start with this "most complicated" API, then go through some real use case examples and see how it fits and if it can be simplified. That should also give us an idea of what could be the best format for results; we initially considered epoch numbers for all, then we realised that for the cyclic case it wasn't the most convenient, then I thought that it wasn't either for the other cases... but I'd like to experiment with a full stack (client CARTO VL expressions and Maps API) before taking a decision.

And thank you very much mate for your comments and examples, you've really know this stuff :heart:

**Comment by Algunenano** [comment]
> On second thought the names are confusing, maybe we could have granularity_unitsand number_or_units or something.

I feel that the naming is a bit confusing but I guess I'm not the only one. I would look for something that reads as `hours of the day` and avoid `granularity` or `multiplicity` names:

Maybe:
```
dimensions: {
  dow: {
    column: 'the_date'
    group_by_quantity: 1,
    group_by_units: 'hours',
    group_by_of: 'day'
  }
}
```

Or even more simple:
```json
dimensions: {
  dow: {
    column: 'the_date'
    group_by: `hourOfDay`
  }
}
```

And just have a list of valid `group_by` names and it's translation to other units. Then, if you decide to open it more (5 minutes of the hour) you can either add an extra string `5minuteOfHour` or add an extra option:
```
dimensions: {
  dow: {
    column: 'the_date'
    group_by: `minuteOfHour`,
    step: 5
  }
}
```

**Comment by dgaubert** [comment]
> Or even more simple:
```js
dimensions: {
  dow: {
    column: 'the_date'
    group_by: 'hourOfDay'
  }
}
```

That was what I had in my mind.

**Comment by jgoizueta** [comment]
Current situation (WIP):

I'm considering some changes to the current implementation, some minor ones/easy to implement:

* Review parameter names
* Solve time-dimensions / dates_as_numbers dependency
* Add dimension filters

And a more complex one:

* Add dimension metadata

## Background

### Tiler API parameters

I'm using this parameters currently for the `dimensions` of the `aggregation`:

```javascript
{
  column: 'the_date',
  groub_by: 'day', // 'dayOfMonth', ...
  group_by_count: 3, // optional (1)
  offset: 0, // optional (0)
  timezone: 'Europe/Madrid' // optional ('utc' == '+0000')
}
```

As mentioned before, `offset` might be confusing because it is used as a time zone offset in the histrograms.
I'm accepting in `timezone` both fixed TZ offsets in seconds and TZ names (including DST rules) accepted by PostgreSQL (so all IANA names are accepted).

The parameters of the internal tiler functions are slightly different:

```javascript
{
    time: `to_timestamp("the_date_epoch")`,
    timezone: 'Europe/Madrid',
    granularity: 'day',
    multiplicity: 1,
    cycle: 'month',
    offset: 0
}
```

### Dimension filtering

We should add dimension filtering to the Maps API (like aggregation columns filter); so it is applied at the proper place of the aggregation query (before/at grouping).

### Dates as numbers

`dates_as_numbers` must be set to true for a layer to be able to use time dimensions. So we should either fail with an error otherwise or automatically set the options when time dimensions are used.

Another option is to adapt the time GROUP BY expressions depending on the value of `dates_as_numbers`, but we'd need to pass all the layer options to the aggregation query function.

### Dimension metadata

As discussed [here](https://github.com/CartoDB/carto-vl/issues/551#issuecomment-426194093) it could be convenient to provide metadata for the grouped dimensions. Unlike for aggregated columns stats, this is not zoom-dependent and can be computed on the base table. But we need share some information between the aggregation query generation and the stats computation.

**Comment by jgoizueta** [comment]
Hey @Algunenano comments are welcome I've been using the PR #1039 with CARTO VL and I'm happy with it. As it is you it gives either numbers (counters of the grouping periods like months or 10-day spans) or ISO strings also (but with some restrictions: no multiple units nor starting time). For day of the week etc no ISO format is supported.

The parameters look like this: 
```javascript
{
    "dimensions": {
        "time_in_days": { // Name of the resulting dimension column
            "column": "the_time", // column discretized for the dimension
            "group": { // defines how to group the values of the column
                "units": "day", // grouping units
                "count": 1, // optional, number of grouping units per period, 1 by default
                "timezone": "Europe/Madrid", // optional (by default 'utc' == '+0000')
                "starting": "2018-01-01" // by default '0001-01-01T00:00:00', periods counted from this time
            },
            "format": "number" // optional, "iso" indicates ISO 8601 format
        }
    }
}
```

**Comment by Algunenano** [comment]
Quick comment:
> // by default '0001-01-01T00:00:00', periods counted from this time

I think it should be 1970 as every other mayor date system

**Comment by Algunenano** [comment]
This was merged in https://github.com/CartoDB/Windshaft-cartodb/pull/1039
